meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java

```diff
@@ -239,6 +240,29 @@ public static MetaClientException translateZkExceptionToMetaclientException(ZkEx
     return new MetaClientException(e);
   }
 
+  public static MetaClientInterface.ConnectState translateKeeperStateToMetaClientConnectState(
+      Watcher.Event.KeeperState keeperState) {
+    if (keeperState == null)
+      return MetaClientInterface.ConnectState.NOT_CONNECTED;
+    switch (keeperState) {
+      case AuthFailed:
+        return MetaClientInterface.ConnectState.AUTH_FAILED;
+      case Closed:
+        return MetaClientInterface.ConnectState.CLOSED_BY_CLIENT;
+      case Disconnected:
+        return MetaClientInterface.ConnectState.DISCONNECTED;
+      case Expired:
+        return MetaClientInterface.ConnectState.EXPIRED;
+      case SaslAuthenticated:
```

Comment:
neat : this is not entirely true. RIght now we don't use ZK SASL authentication so you will always see Sasl auth failure. But still it's not "auth failed" case because we do x509 authentication. Right not it's fine but may be you can add TODO to fix this when we want to onboard on ZK ACL.

Reply:
When we dont use ZK SASL authentication, will Zk Server return auth failed?

Reply:
we never use sasl authentication and hence you will always see error log about "sasl auth failure." Whenever x509/any other authentication fails then zk server will return auth failed.

Reply:
Had a discussion offline, for any authentication scheme being used by underlying zk client, it should return "AUTHENTICATED->CONNECTED" events in this order and in case of auth failure it should be "AUTH_FAILED" event.  Now if it doesn't use SASL then it should not return AUTH_FAILED event.